crates/rbuilder/src/provider/ipc_state_provider.rs

```diff
@@ -0,0 +1,582 @@
+}
+
+impl IpcStateProvider {
+    /// Crates new instance of state provider
+    fn new(
+        ipc_provider: RpcProvider,
+        block_id: BlockId,
+        code_cache: Arc<DashMap<B256, Bytecode>>,
+    ) -> Self {
+        Self {
+            ipc_provider,
+            block_id,
+
+            code_cache,
+
+            block_hash_cache: DashMap::new(),
+            storage_cache: DashMap::new(),
+            account_cache: DashMap::new(),
+        }
+    }
+
+    /// Crates new instance of state provider on the heap
+    // Box::new(Arc::new(Self)) is required because StateProviderFactory returns Box<dyn StateProvider>
+    #[allow(clippy::redundant_allocation)]
+    fn into_boxed(
+        ipc_provider: RpcProvider,
+        block_id: BlockId,
+        code_cache: Arc<DashMap<B256, Bytecode>>,
+    ) -> Box<Arc<Self>> {
+        Box::new(Arc::new(Self::new(ipc_provider, block_id, code_cache)))
+    }
+}
+
+impl StateProvider for IpcStateProvider {
+    /// Get storage of given account
+    fn storage(
+        &self,
+        account: Address,
+        storage_key: StorageKey,
+    ) -> ProviderResult<Option<StorageValue>> {
+        let id: u64 = rand::random();
+        let span = trace_span!("storage", id);
+        let _guard = span.enter();
+        trace!("storage:get");
+
+        if let Some(storage) = self.storage_cache.get(&(account, storage_key)) {
+            return Ok(Some(*storage));
+        }
+
+        let key: U256 = storage_key.into();
+        let storage = self
+            .ipc_provider
+            .call::<_, StorageValue>("eth_getStorageAt", (account, key))
+            .map_err(ipc_to_provider_error)?;
+
+        self.storage_cache.insert((account, storage_key), storage);
+
+        trace!("got storage");
+        Ok(Some(storage))
+    }
+
+    /// Get account code by its hash
+    /// IMPORTANT: Assumes remote provider (node) has RPC call:"rbuilder_getCodeByHash"
+    fn bytecode_by_hash(&self, code_hash: &B256) -> ProviderResult<Option<Bytecode>> {
+        let id: u64 = rand::random();
+        let span = trace_span!("bytecode", id);
+        let _guard = span.enter();
+        trace!("bytecode:get");
+
+        let empty_hash = code_hash.is_zero() || *code_hash == KECCAK_EMPTY;
+        if empty_hash {
+            return Ok(None);
+        }
+
+        if let Some(bytecode) = self.code_cache.get(code_hash) {
+            return Ok(Some(bytecode.clone()));
+        }
+
+        let bytes = self
+            .ipc_provider
+            .call::<_, Bytes>("rbuilder_getCodeByHash", (code_hash,))
+            .map_err(ipc_to_provider_error)?;
+
+        let bytecode = Bytecode::new_raw(bytes);
+
+        self.code_cache.insert(*code_hash, bytecode.clone());
+        trace!("bytecode: got");
+
+        Ok(Some(bytecode))
+    }
+}
+
+impl BlockHashReader for IpcStateProvider {
+    /// Get the hash of the block with the given number. Returns `None` if no block with this number exists
```

Comment:
So this function will never return None, because we wrap this case in JsonErrPayloadMisinterpretedAsSuccess?
If it's the case we should match on this error in here and return None